### ISSUE TITLE
Rename `contextProps` to `pageContext`.

### ISSUE THREAD

**Issue by brillout** [brillout/vite-plugin-ssr#58]

**Comment by brillout** [comment]
This is the last breaking change before I release the first non-beta version.

I will provide scripts to ease migration, e.g. `grep addContextProps | xargs sed -i 's/addContextProps/addSsrContext/g'`.

Let me know if you object with the new name `ssrContext`. I actually feel like there could be a better name but I can't find one.

@deckchairlabs 
@doeixd 
@gryphonmyers 
@chrisvariety

(@gryphonmyers I will take care of updating the PR and resolve merge confits.)

**Comment by brillout** [comment]
cc @ozergul @alsyapukov @aral @CanRau @ruslankonev @naranjamecanica

**Comment by rulentor** [comment]
Hi!
doc here (https://github.com/brillout/vite-plugin-ssr) not true for boilerplate template.
How me get data from (https://swapi.dev/api/films/${movieId}/?format=json) by node-fetch and out Page as json. How many times knocking result only empty {} in browser (link /movies/1) How knew clearing mind truth? Big fans before

**Comment by brillout** [comment]
@rulentor write me on Discord

**Comment by deckchairlabs** [comment]
Why can't it just be simply `context`?

**Comment by brillout** [comment]
@deckchairlabs Because it will conflict with other `context` objects, e.g. https://github.com/brillout/wildcard-api. Imagine a framework built on top of vite-plugin-ssr and Wildcard API; it will be confusing to have several `context` objects that mean different things.

**Comment by brillout** [comment]
I'm thinking to name it `renderContext`.

The thing is that `contextProps` lives *only* during rendering; the `contextProps` object is created when a page rendering starts (it's actually the user that creates the object when he does `await renderPage({ url, contextProps: {} })` and `contextProps` is (usually) free for garbage collection after the page rendering is done.

`renderContext` may sound weird at first but it might actually clarify things. I'll meditate about it.

**Comment by deckchairlabs** [comment]
I see, how about something more event/hook based? `onBeforeRender`

```ts
// _default.server.page.ts
export async function onBeforeRender({ contextProps, Page }) {
    return {
        contextProps: {
            ...contextProps,
            foo: 'bar'
        }
    }
}
```

**Comment by brillout** [comment]
Interesting, I didn't think about this.

**Comment by deckchairlabs** [comment]
I'm not sure if I've mentioned the below before on a different thread, apologies if I have.

This may be out of scope, but I've been thinking if the naming, especially for the server API, was more focused around what `stage` the render cycle was in.

```ts
// _default.server.page.ts
// Could redirect
export async function onRequest() {
  // Send a redirect perhaps, unauthorized etc?
  // Return nothing just continues with render?
}

export async function onResponse(renderResult: string, originalRequest: Request) {
  // Could modify the HTML perhaps?
  return renderResult.
}

export async function onBeforeRender({ contextProps, Page }) {
    return {
        contextProps: {
            ...contextProps,
            foo: 'bar'
        }
    }
}
```

**Comment by brillout** [comment]
@deckchairlabs I like your idea. I'll medidate about renaming `addContextProps` to `onBeforeRender`. And good point that it will make it natural to add further hooks shall the need arise.

**Comment by chrisvariety** [comment]
`addContextProps` is just half of the story though, what are we calling it in `function render({ Page, contextProps }: { Page: ReactComponent; contextProps: ContextProps }) { ... }` ? It seems a little weird to rename to `onBeforeRender` but still have the render function have `contextProps`.

**Comment by brillout** [comment]
@chrisvariety Yes the idea is to rename *all* `contextProps` variables.

For example, renaming all `contextProps` to `renderContext` while also renaming `addContextProps` to `onBeforeRender`.

```js
export async function onBeforeRender({ renderContext, Page }) {
    return {
        // Add stuff to `renderContext`
        renderContext: {
            foo: 'bar'
        }
    }
}

export async function render({ renderContext, Page }) {
   // ...
}
```

**Comment by CanRau** [comment]
Think I'm fine with all the suggesting, thanks for pinging and asking 🙌

🤜🤛

**Comment by gryphonmyers** [comment]
This all sounds good, however I'm now even more confused by the prospect of doing client-side fetching as described in https://github.com/brillout/vite-plugin-ssr/issues/52. If they're now `renderProps`, how / why does the client have the ability to affect them?

**Comment by brillout** [comment]
@gryphonmyers

I'm still pondering the renaming.

So far I like the `renderContext` name as it makes it clearer that the `renderContext` object lives only for the purpose of rendering.

I'm hesitating about `onBeforeRender` precisely because, when using `useClientRouter()` and when the user navigates to a new page, then it's somewhat confusing that the only thing called on the server is `onBeforeRender`.

**Comment by brillout** [comment]
From discord https://discord.com/channels/815937377888632913/815937377888632916/842824951350427688

> `pageContext` and `getPageContext` is what came to my mind. Since it is all about the page (no?).

My reply:

> I like your idea, makes sense. It somewhat conflicts with `pageProps` but we can rename that as well.

**Comment by brillout** [comment]
My current favorite is now `pageContext`.

**Comment by brillout** [comment]
I'm leaning towards renaming `addContextProps` to `addPageContext`.

I do like the `onBeforeRender` idea, but it's sightly misleading for people who use `useClientRouter()` (with Client-side Routing, when the user navigates to a new page, the browser makes a server request and `onBeforeRender()` is called while the `render()` hook is *not* called: the new page is *only* rendered to the DOM and is not rendered to HTML).

Happy to be contradicted and discuss.

**Comment by brillout** [comment]
I decided for `pageContext`.

I'm still hesitating about the hook thing.

**Comment by brillout** [comment]
I'm going to rename `contextProps` to `pageContext` and `addContextProps` to `addPageContext` in the next couple of days.

I'm still thinking about the stage-hook thing. I like it but I still need time to let the idea sink in. If anyone as more thoughts about stage-hooks, now is the time.

Looking forward to see what we come up with. Well designed stage-hooks could be a wonderful thing.

Opened a new ticket for it https://github.com/brillout/vite-plugin-ssr/issues/71.

**Comment by gryphonmyers** [comment]
> I'm going to rename `contextProps` to `pageContext` and `addContextProps` to `addPageContext` in the next couple of days.
> 
> I'm still thinking about the stage-hook thing. I like it but I still need time to let the idea sink in. If anyone as more thoughts about stage-hooks, now is the time.
> 
> Looking forward to see what we come up with. Well designed stage-hooks could be a wonderful thing.
> 
> Opened a new ticket for it #71.

I guess my main thought there is if the stage hooks are under consideration and hypothetically do get implemented, would `addPageContext` be better suited as an `on*` hook for consistency?

**Comment by brillout** [comment]
Yes, once/if we go with stage-hooks, the plan is to then deprecate `addPageContext`.

**Comment by brillout** [comment]
> I'm going to rename `contextProps` to `pageContext` and `addContextProps` to `addPageContext` in the next couple of days.

Done and released in `0.1.0-beta.45`.

Stage hooks are still open for consideration at #71.

**Comment by brillout** [comment]
Every hook is now a stage hook. I.e. `addPageContext()` is now `onBeforeRender()`.

Thanks for all the lovely feedback.
